### ISSUE TITLE
[feature request] keep last n built assets

### ISSUE THREAD

**Issue by plrthink** [mipearson/webpack-rails#48]
First of all, I love this gem, it helps me to integrate webpack with rails without plain. With a little customization it works so greate in both environment.

One thing I think it's missing is that it can't preserve last n builds between deployments like sprockets-rails does. I think it's essential because we not only want to keep assets during deployment, but remove older ones to not eat up our disk.

**Comment by mipearson** [comment]
As in keep the manifests as well as the compiled assets? There's nothing in the compile step that would remove the old assets, but the manifest is overwritten.

Thank you for your kind words regarding the gem :)

**Comment by plrthink** [comment]
I might describe the feature in the wrong way. I mean only keep last n time builds and remove the older ones. sprockets-rails has this feature as describes [here](https://github.com/rails/sprockets-rails#rake-task). If we don't remove older ways, the size of assets folder will get larger and larger.

**Comment by mipearson** [comment]
Oh, right! 

In our case the assets we compile with webpack are quite small - JS only, no images or CSS - so it hasn't hit us yet.

I would need to look at how sprockets does it, but it'd be a good thing to support.

**Comment by plrthink** [comment]
Yeah, I would also want to figure out and contribute to this gem.

**Comment by plrthink** [comment]
I have some time to search on sprockets codebase. I find [this method](https://github.com/rails/sprockets/blob/9db6aa1c8bec0c445c58d2783d7166f7c44a58f0/lib/sprockets/manifest.rb#L248), it describes the algorithm about cleaning older versions of assets. I think we can use it. 

However, there's one different thing. Webpack seems like don't output the information about when the asset file is built like Sprockets does. But it's fine, we can check the `mtime` on the file itself.

**Comment by plrthink** [comment]
Oh, I miss that mainfest.json emitted by sprockets also have info about every versions of assets like below:

```` json
"workspace-901e9d31f7757378882ea086cf3c65a7.js": {
  "logical_path": "workspace.js",
  "mtime": "2016-06-17T21:01:07+08:00",
  "size": 892279,
  "digest": "901e9d31f7757378882ea086cf3c65a7"
},
"workspace-408babeced685f1dc6d1ec68806cca81.css": {
  "logical_path": "workspace.css",
  "mtime": "2016-07-27T11:41:27+08:00",
  "size": 195125,
  "digest": "408babeced685f1dc6d1ec68806cca81"
},
"workspace-16ed83e3e9615b664a8d8113dc54222d.js": {
  "logical_path": "workspace.js",
  "mtime": "2016-08-07T15:32:02+08:00",
  "size": 916000,
  "digest": "16ed83e3e9615b664a8d8113dc54222d"
},
"workspace-1eef72f05291133f14d3322772d2db0b.css": {
  "logical_path": "workspace.css",
  "mtime": "2016-08-07T15:13:32+08:00",
  "size": 195260,
  "digest": "1eef72f05291133f14d3322772d2db0b"
}
````

It helps to find all previous version of files. So I think instead of overwriting `manifest.json` every time, we should mutate that to include new assets'info and also remove older ones like sprockets does.

**Comment by mipearson** [comment]
Harder than it sounds, unfortunately - dig in to how webpack is generating the manifest and you'll see what I mean.

**Comment by plrthink** [comment]
sounds like we need help here

**Comment by mipearson** [comment]
I would recommend a post-build step that parses the manifest.json that is created and stores a separate history file, which can then be read.

**Comment by focused** [comment]
Any progress on this? Very useful gem and I need this feature too.

**Comment by ujh** [comment]
Wouldn't it be enough to delete the old version after the deployment, i.e. remove all files not listed in the manifest? Obviously it would have to be a separate task that's executed one the server is restarted.

**Comment by focused** [comment]
Yes, it should be performed after cleaning old releases, also it would be great to compile webpack assets locally (otherwise server cpu load increased dramatically).

**Comment by ujh** [comment]
Those seem like two different tasks to me. If we have a task that removes the old data and you execute it after you copy your files to the server then your use case should be covered as well, shouldn't it?

**Comment by focused** [comment]
Yes, these are different tasks. If you keep two or more releases, you can remove old data safely, but if you keep only one release, you'll probably see your web application without css/js for a second while puma/unicorn restarts. I think it's safer to remove old assets just after restart event.

**Comment by plrthink** [comment]
I'm kind of lost in there. What's the purpose of keep last n releases? I thought it is to ensure user won't get pages without css/js while deployments, but after reading @focused 's comments I think I'm misunderstanding this.
